tjmonopix2/scans/eudaq_producer.py

```diff
@@ -0,0 +1,490 @@
+            glitch_detected = False
+            error = False
+            # Split can return empty data, thus do not return send empty data
+            # Otherwise fragile EUDAQ will fail. It is based on very simple event counting only
+
+            if np.any(dat):
+                trigger = dat[0] & au.TRG_MASK
+
+                if np.bitwise_and(dat[0], au.TRIGGER_HEADER) == 0:
+                    # this might happen the very first time we get data
+                    # if there is trash data in the FIFO (which does not even start with a trigger word)
+                    # we would interpret this trash 'normal' data word as a trigger word
+                    # skip this stuff
+                    continue
+
+                prev_trg = self.last_trigger
+                if self.last_trigger > 0 and trigger != self.last_trigger + 1:    # Trigger number jump
+                    error = True
+                    if (self.last_trigger + 1) == (trigger >> 1):
+                        # Measured trigger number is exactly shifted by 1 bit, due to glitch
+                        glitch_detected = True
+
+
+                self.last_trigger = trigger if not glitch_detected else (trigger >> 1)
+
+
+                if self.last_trigger < 100 and prev_trg > 32760:  # trigger number overflow
+                    # arbitrary choice of borders, in case we missed triggers do not use a '=='
+                    print('looks like an ovrFlw')
+                    self.ovflw_cnt += 1
+                elif error and not glitch_detected:
+                    self.log.warning('Expected != Measured trigger number: %d != %d',
+                                     self.last_trigger + 1, trigger)
+
+                # print('sending event with trgNmb = ', self.last_trigger)
+                data_to_send = np.concatenate(trigger_data[i:i + 2])
+
+
+                # The event building works in the following way
+                # Data from FIFO eg:
+                # D D D T0 D D D D T1 D D D D D T2 D D D T3 D D D D
+                #       |<---0--->||<----1---->||<--2-->||<-------... (store for next execution)
+                # D ... data word, Tx ... Trigger word, |<-x->| ... EUDAQ event with trigger number x
+                # investigation of the Chip efficiency in a testbeam setup proved that when doing the event building in
+                # the way scetched above, some pixel hits got assigned with a wrong trigger number. Eg:
+                # D after T2 would actually still belong to T1 (shown by reconstruction with telescope setup)
+                # as an easy workaround we assign all the data to 2 trigger numbers
+                # with the scetch from above this looks like:
+                #  D D D T0 D D D D T1 D D D D D T2 D D D T3 D D D D
+                #        |<----------0---------->|
+                #                   |<--------1--------->|
+                #                                |<---------------------... next execution
```

Comment:
This is something we should talk about and probably test during testbeam. With a proper veto length, triggers are vetoed in the TLU until all hit words are read out (effectively limiting the trigger rate, but to a very high value). 
Hit duplication opens a lot of room for mistakes.